### PR TITLE
CPU warp, with tests.

### PR DESCRIPTION
**Pull request by mzient**
Signed-off-by: Michal Zientkiewicz <michalz@nvidia.com>

#### Why we need this PR?
*Pick one*
- It adds WarpCPU kernel to use in Warp operator family.

#### What happened in this PR?
 - What was changed, added, removed?
  * Added WarpCPU kernel class.
  * Moved some utilities outside files used by GPU implementation.
 - What is most important part that reviewers should focus on?
  * `warp_cpu.h`
 - Was this PR tested?
  * Unit tests
 - Were docs and examples updated, if necessary?
  * No

**JIRA TASK**: [DALI-826]

### CODE REVIEW

**Comment by mzient** [comment]
!build

File: dali/kernels/imgproc/warp_cpu.h [hunk 1]
```diff
@@ -0,0 +1,118 @@
...
+      KernelContext &context,
+      const InTensorCPU<InputType, tensor_ndim> &input,
+      const MappingParams &mapping_params,
+      const TensorShape<spatial_ndim> &out_size,
+      DALIInterpType interp = DALI_INTERP_LINEAR,
+      const BorderType &border = {}) {
+    KernelRequirements req;
+    req.output_shapes.resize(1);
+    auto out_shape = shape_cat(out_size, input.shape[channel_dim]);
+    req.output_shapes[0] = TensorListShape<tensor_ndim>({out_shape});
```
Comment (jantonguirao) [inline dali/kernels/imgproc/warp_cpu.h]:
```suggestion
    req.output_shapes.push_back(TensorListShape<tensor_ndim>({out_shape}));
```
and remove the resize(1) call?

File: dali/kernels/imgproc/warp_cpu.h [hunk 2]
```diff
@@ -0,0 +1,118 @@
...
+#include "dali/kernels/kernel.h"
+#include "dali/kernels/imgproc/warp/mapping_traits.h"
+#include "dali/kernels/imgproc/sampler.h"
+#include "dali/kernels/imgproc/warp/map_coords.h"
+
+namespace dali {
+namespace kernels {
+
+/// @remarks Assume HWC layout
+template <typename _Mapping, int ndim, typename _OutputType, typename _InputType,
```
Comment (jantonguirao) [inline dali/kernels/imgproc/warp_cpu.h]: Also, I think we agreed in previous kernels to use an order of template arguments like: `typename OutputType, typename InputType, int ndim`

File: dali/kernels/test/warp_test/warp_cpu_test.cc [hunk 3]
```diff
@@ -0,0 +1,150 @@
+// Copyright (c) 2019, NVIDIA CORPORATION. All rights reserved.
+//
```
Comment (jantonguirao) [inline dali/kernels/test/warp_test/warp_cpu_test.cc]: I mentioned it before, but here it goes again: this `warp_test` directory is not consistent with what we do in the rest of DALI (keeping `*_test.*` files in the same directory as the implementation files).

File: dali/kernels/imgproc/warp_cpu.h [hunk 4]
```diff
@@ -0,0 +1,118 @@
...
+  KernelRequirements Setup(
+      KernelContext &context,
+      const InTensorCPU<InputType, tensor_ndim> &input,
+      const MappingParams &mapping_params,
+      const TensorShape<spatial_ndim> &out_size,
+      DALIInterpType interp = DALI_INTERP_LINEAR,
+      const BorderType &border = {}) {
+    KernelRequirements req;
+    req.output_shapes.resize(1);
+    auto out_shape = shape_cat(out_size, input.shape[channel_dim]);
```
Comment (JanuszL) [inline dali/kernels/imgproc/warp_cpu.h]: How you want to handle a case when the out_shape is calculated based on the transformation (so output canvas is a big as it needs to house the transformed image).

**Comment by JanuszL** [comment]
Please adjust comments `///` -> `/* */`.

File: dali/kernels/imgproc/warp_cpu.h [hunk 5]
```diff
@@ -0,0 +1,118 @@
...
+/// @remarks Assume HWC layout
+template <typename _Mapping, int ndim, typename _OutputType, typename _InputType,
+          typename _BorderType>
+class WarpCPU {
+ public:
+  static constexpr int spatial_ndim = ndim;
+  static constexpr int tensor_ndim = ndim + 1;
+  static constexpr int channel_dim = ndim;
+
+  static_assert(spatial_ndim == 2, "Not implemented for spatial_ndim != 2");
```
Comment (szalpal) [inline dali/kernels/imgproc/warp_cpu.h]:
```suggestion
  static_assert(spatial_ndim == 2, "This kernel's implementation assumes HWC layout, therefore spatial_ndim must be equal to 2");
```
It's not that this kernel is not implemented for `spatial_ndim!=2`, but it **is implemented for `spatial_ndim==2`**. Logically there's no difference, but semantically there is...

File: dali/kernels/test/warp_test/warp_cpu_test.cc [hunk 6]
```diff
@@ -0,0 +1,150 @@
...
+    { 1, 0, 0 }
+  }};
+
+  cv::Mat cv_img = cv::imread(testing::dali_extra_path() + "/db/imgproc/alley.png");
+  auto cpu_img = view_as_tensor<uint8_t>(cv_img);
+
+  WarpCPU<AffineMapping2D, 2, uint8_t, uint8_t, BorderClamp> warp;
+
+  ScratchpadAllocator scratch_alloc;
+
```
Comment (szalpal) [inline dali/kernels/test/warp_test/warp_cpu_test.cc]: I got confused here. On one hand you use `Scratchpad*` here, on the other hand it doesn't seem to be used in `WarpCPU` class.

File: dali/kernels/test/warp_test/warp_cpu_test.cc [hunk 7]
```diff
@@ -0,0 +1,150 @@
...
+namespace dali {
+namespace kernels {
+
+inline void IsWarpKernelValid() {
+  check_kernel<WarpCPU<AffineMapping2D, 2, float, uint8_t, float>>();
+}
+
+TEST(WarpCPU, Affine_Transpose_Single) {
+  AffineMapping2D mapping_cpu = mat2x3{{
+    { 0, 1, 0 },
```
Comment (szalpal) [inline dali/kernels/test/warp_test/warp_cpu_test.cc]: Could you bring out the duplicated code for setting up these tests to fixture-class?
Then it would be possible to test it against different input-output pairs using `INPUT_OUTPUT_TYPED_TEST_SUITE` and code duplication will disappear

File: dali/kernels/test/warp_test/warp_cpu_test.cc [hunk 8]
```diff
@@ -0,0 +1,150 @@
...
+#include "dali/test/dali_test_config.h"
+#include "dali/core/geom/transform.h"
+#include "dali/kernels/test/warp_test/warp_test_helper.h"
+
+namespace dali {
+namespace kernels {
+
+inline void IsWarpKernelValid() {
+  check_kernel<WarpCPU<AffineMapping2D, 2, float, uint8_t, float>>();
+}
```
Comment (szalpal) [inline dali/kernels/test/warp_test/warp_cpu_test.cc]: In my opinion, it would be more readable, if this was actually a test:
```
TEST(WarpCPU, check_kernel_compile_time) {
  check_kernel<WarpCPU<AffineMapping2D, 2, float, uint8_t, float>>();
  SUCCEED(); // I'd put it here, to make it clear, that this tests passes if compiles.
}
```

File: dali/kernels/test/warp_test/warp_cpu_test.cc [hunk 9]
```diff
@@ -0,0 +1,150 @@
...
+#include "dali/kernels/test/test_tensors.h"
+#include "dali/kernels/scratch.h"
+#include "dali/kernels/alloc.h"
+#include "dali/test/dali_test_config.h"
+#include "dali/core/geom/transform.h"
+#include "dali/kernels/test/warp_test/warp_test_helper.h"
+
+namespace dali {
+namespace kernels {
+
```
Comment (szalpal) [inline dali/kernels/test/warp_test/warp_cpu_test.cc]: I'd add a namespace for testing, especially that there are some helper-functions here.  Moreover, I'd add namespace `warp` to emphasize, that this code is for warp actually...

File: dali/kernels/imgproc/warp_cpu.h [hunk 10]
```diff
@@ -0,0 +1,118 @@
...
+
+    VALUE_SWITCH(interp, static_interp, (DALI_INTERP_NN, DALI_INTERP_LINEAR),
+      (RunImpl<static_interp>(context, output, input, mapping, border);),
+      (DALI_FAIL("Unsupported interpolation type"))
+    ); // NOLINT
+  }
+
+  template <DALIInterpType static_interp>
+  void RunImpl(
+      KernelContext &context,
```
Comment (szalpal) [inline dali/kernels/imgproc/warp_cpu.h]: I guess this function could be brought out:
```
namespace warp {
void RunImpl();
}
```

Plus I think it deserves its own unit test. The thing is, that right now we can assume, that it's tested 1:1 with `Run` itself, however in future somebody might add something to `Run` and create a case, where `RunImpl` wouldn't be as tightly coupled and tested as it is now

File: dali/kernels/test/warp_test/warp_gpu_test.cu [hunk 11]
```diff
@@ -41,7 +41,7 @@ class WarpPrivateTest {
   }
 };
 
-void IsWarpKernelValid() {
+inline void IsWarpKernelValid() {
```
Comment (szalpal) [inline dali/kernels/test/warp_test/warp_gpu_test.cu]: Same as before, maybe it'd be better to actually make it a TEST?

File: dali/kernels/imgproc/warp/map_coords.h [hunk 12]
```diff
@@ -0,0 +1,49 @@
...
+  // Since ultimately our destination coordinates are integer, we need to move to pixel-center
+  // frame before applying the mapping and passing the output to a sampler.
+  return m(pos + 0.5f);
+}
+
+template <typename Mapping, std::size_t dim>
+DALI_HOST_DEV
+enable_if_t<!is_fp_mapping<Mapping>::value, ivec<dim>> map_coords(const Mapping &m, ivec<dim> pos) {
+  // When given integer point coordinates, samplers simply uses them as 0-based indices.
+  // If the mapping producces integral coordinates, there's no point in going from indices to
```
[the remark this answers is not in the document]
Comment (mzient) [reply dali/kernels/imgproc/warp/map_coords.h]: It does apply.

File: dali/kernels/imgproc/warp_cpu.h (hunk 2 above)
Comment (mzient) [reply dali/kernels/imgproc/warp_cpu.h]: Did we? I think it was a rather unilateral opinion.

File: dali/kernels/imgproc/warp_cpu.h (hunk 5 above)
Comment (mzient) [reply dali/kernels/imgproc/warp_cpu.h]: 1. This is _temporary_. It it wasn't, the ndim parameter would even be here.
2. Your proposed message is misleading. The limitation on number of dimensions is not a consequence of layout constraint, but lack of "Surface3D" and a corresponding sampler.
I can rephrase it to "This kernel is implemented only for 2D data." - however, the phrase "Not implemented" usually carries an additional message which reads "not implemented *yet*". Stating that the implementation assumes something doesn't suggest that it was ever meant to support something else.

File: dali/kernels/imgproc/warp_cpu.h (hunk 4 above)
Comment (mzient) [reply dali/kernels/imgproc/warp_cpu.h]: It's the responsibility of the operator, not the kernel.

File: dali/kernels/imgproc/warp_cpu.h (hunk 10 above)
Comment (mzient) [reply dali/kernels/imgproc/warp_cpu.h]: I'd go the other way and make it private.

File: dali/kernels/test/warp_test/warp_cpu_test.cc (hunk 7 above)
Comment (mzient) [reply dali/kernels/test/warp_test/warp_cpu_test.cc]: I don't want to test Cartesian product of types here. I don't want to expose it in operator, either - as such, INPUT_OUTPUT_TYPED_TEST_SUITE is inadequate.

File: dali/kernels/test/warp_test/warp_cpu_test.cc (hunk 6 above)
Comment (mzient) [reply dali/kernels/test/warp_test/warp_cpu_test.cc]: Handling scratchpad memory is part of the Kernel API contract. I know it's not used _now_, but kernels should _never_ be used without a valid scratchpad - otherwise the program is ill-formed.

File: dali/kernels/imgproc/warp_cpu.h (hunk 10 above)
Comment (szalpal) [reply dali/kernels/imgproc/warp_cpu.h]: What's the benefit of making it private? It doesn't use any member of kernel and you just want to dodge of writing unit-test for this :p

File: dali/kernels/test/warp_test/warp_cpu_test.cc (hunk 7 above)
Comment (szalpal) [reply dali/kernels/test/warp_test/warp_cpu_test.cc]: Still you have bunch of duplicated code there. Plus, since this kernel should process variety of input-output type pair, there will be more (e.g. for `float16`, which test here also is lacking...)

File: dali/kernels/test/warp_test/warp_cpu_test.cc (hunk 6 above)
Comment (szalpal) [reply dali/kernels/test/warp_test/warp_cpu_test.cc]: > kernels should never be used without a valid scratchpad

So, there's nothing about this requirement in the blueprints, so I assume it's not a requirement. Test not only verifies the code, but also documents it: creates an example of usage. With this here used gets confused.

So, IMHO, either let's enforce using `Scratchpad` in API (like the `check_kernel` or sth) or remove it from here. Just to be consistent

File: dali/kernels/imgproc/warp_cpu.h (hunk 5 above)
Comment (szalpal) [reply dali/kernels/imgproc/warp_cpu.h]: I'd see no point of making the message suggest, that at some point in the future (not really clearly defined) we will add some functionality, because:
1. It's not set up, when we do it
2. When we decide "ok, we may do it" this can be no longer needed

My message is just a suggestion, where I wanted to say, that this kernel is actually implemented only regarding to 3-dimension (including channels) image, which serves as a remark.

Comment (mzient) [reply dali/kernels/imgproc/warp_cpu.h]: I don't want to mix layout with number of spatial dimensions. It's just confusing.

File: dali/kernels/imgproc/warp_cpu.h (hunk 10 above)
Comment (mzient) [reply dali/kernels/imgproc/warp_cpu.h]: One benefit is avoiding double validation of input parameters.

File: dali/kernels/test/warp_test/warp_cpu_test.cc (hunk 9 above)
Comment (mzient) [reply dali/kernels/test/warp_test/warp_cpu_test.cc]: I think that namespace `dali::testing` should be used for testing tools, not for _unit_ tests - these should be close to the definition of the functionality being tested to avoid unnecessary qualified names. Putting `testing` inside `dali::kernels` or `dali::kernels::warp` creates confusion with `dali::testing'.

File: dali/kernels/test/warp_test/warp_cpu_test.cc (hunk 8 above)
Comment (mzient) [reply dali/kernels/test/warp_test/warp_cpu_test.cc]: Fair enough.

File: dali/kernels/test/warp_test/warp_cpu_test.cc (hunk 6 above)
Comment (mzient) [reply dali/kernels/test/warp_test/warp_cpu_test.cc]: From the documentation of Run in `kernel.h`:
```
   * @param context - environment; provides scratch memory, cuda stream, batch info, etc.
   *                  Scratch area must satisfy requirements returned by Setup.
```

**Comment by mzient** [comment]
!build

File: dali/kernels/imgproc/warp_gpu.h [hunk 13]
```diff
@@ -27,15 +27,23 @@ namespace dali {
...
+ * The warping uses a mapping functors to map destination coordinates to source
+ * coordinates and samples the source tensors at the resulting locations.
+ *
+ * @remarks
+ *  * Assumes HWC layout
+ *  * Output and input have same number of spatial dimenions
+ *  * Output and input have same number of channels and layout
  */
-template <typename _Mapping, int ndim, typename _OutputType, typename _InputType,
+template <typename _Mapping, int _spatial_ndim, typename _OutputType, typename _InputType,
```
Comment (jantonguirao) [inline dali/kernels/imgproc/warp_gpu.h]: question: why not `spatial_dim` and get rid of the definition in line 45?

File: dali/kernels/test/warp_test/warp_gpu_test.cu [hunk 14]
```diff
@@ -41,8 +41,9 @@ class WarpPrivateTest {
   }
 };
 
-void IsWarpKernelValid() {
+TEST(WarpGPU, check_kernel) {
   check_kernel<WarpGPU<AffineMapping2D, 2, float, uint8_t, float>>();
+  SUCCEED();
```
Comment (jantonguirao) [inline dali/kernels/test/warp_test/warp_gpu_test.cu]: unless this is testing anything. I'd say remove it

**Comment by mzient** [comment]
!build

File: dali/kernels/test/warp_test/warp_gpu_test.cu (hunk 14 above)
Comment (mzient) [reply dali/kernels/test/warp_test/warp_gpu_test.cu]: That's the way @szalpal wanted it and I'm really fed up with arguing about such details...

File: dali/kernels/imgproc/warp_gpu.h (hunk 13 above)
Comment (mzient) [reply dali/kernels/imgproc/warp_gpu.h]: Because I want it to be accessible as a dependent name `Kernel::spatial_dim`.

File: dali/kernels/test/warp_test/warp_cpu_test.cc (hunk 7 above)
Comment (mzient) [reply dali/kernels/test/warp_test/warp_cpu_test.cc]: I don't see much duplicated code in this particular case, because it has a different validation procedure than the other ones.

Comment (mzient) [reply dali/kernels/test/warp_test/warp_cpu_test.cc]: I tried to to it and it's definitely not shorter - it's a topic for another PR.

File: Doxyfile [hunk 15]
```diff
@@ -796,7 +796,10 @@ INPUT_ENCODING         = UTF-8
 # *.m, *.markdown, *.md, *.mm, *.dox, *.py, *.pyw, *.f90, *.f, *.for, *.tcl,
 # *.vhd, *.vhdl, *.ucf, *.qsf, *.as and *.js.
 
-FILE_PATTERNS          =
+FILE_PATTERNS          = *.c *.cpp *.cc *.C *.cxx *.h *.hpp *.hxx *.H \
+                         *.inc *.inl *.cu *.cuh *.m \
+                         *.idl *.f *.py *.s *.asm *.S *.ptx *.sass *.ll \
+                         *.cmake *.dox *.markdown *.md
```
Comment (klecki) [inline Doxyfile]: Do we really want to generate doxygen from .cmake or .md files?

Comment (mzient) [reply Doxyfile]: Not sure about cmake, but .md is nice to have - doxygen will format it for you.

**Comment by mzient** [comment]
!build

File: Doxyfile (hunk 15 above)
Comment (klecki) [reply Doxyfile]: The thing is we already handle .md (or rather .rst that we use) through sphinx.

File: dali/kernels/test/warp_test/warp_test_helper.h [hunk 16]
```diff
@@ -0,0 +1,40 @@
...
+#define DALI_KERNELS_TEST_WARP_TEST_WARP_TEST_HELPER_H_
+
+#include "dali/kernels/imgproc/warp/affine.h"
+
+namespace dali {
+namespace kernels {
+
+/// @brief Apply correction of pixel centers and convert the mapping to
+///        OpenCV matrix type.
+inline cv::Matx<float, 2, 3> AffineToCV(const AffineMapping2D &mapping) {
```
Comment (klecki) [inline dali/kernels/test/warp_test/warp_test_helper.h]: Should we collect all the OpenCV test helpers in one place?

File: Doxyfile (hunk 15 above)
Comment (mzient) [reply Doxyfile]: .md and .markdown are handled by default anyway - i didn't want to _remove_ anything.

File: dali/kernels/test/warp_test/warp_test_helper.h (hunk 16 above)
Comment (mzient) [reply dali/kernels/test/warp_test/warp_test_helper.h]: I don't know... this one is quite specific - i.e. it applies only to WarpAffine, but I don't think e.g. keypoint transformation should use the same fixup.

File: dali/kernels/imgproc/warp/map_coords.h [hunk 17]
```diff
@@ -0,0 +1,59 @@
...
+namespace dali {
+namespace kernels {
+namespace warp {
+
+/** @brief Converts destination _integer_ coordinates to _floating point_ source coordinates
+ *
+ * This function is used by warp implementations to calculate source coordinates using
+ * using non-integer mapping. The coordinates are converted to pixel-centered.
+ */
+template <typename Mapping, std::size_t dim>
```
Comment (jantonguirao) [inline dali/kernels/imgproc/warp/map_coords.h]: use `int dim`

File: dali/kernels/test/warp_test/warp_gpu_test.cu (hunk 14 above)
Comment (jantonguirao) [reply dali/kernels/test/warp_test/warp_gpu_test.cu]: I convinced him to get rid of it in another PR. Anyway, won't push more on this

File: dali/kernels/imgproc/warp/map_coords.h (hunk 17 above)
Comment (mzient) [reply dali/kernels/imgproc/warp/map_coords.h]: I must use whatever vec is using.

Comment (mzient) [reply dali/kernels/imgproc/warp/map_coords.h]: I'm considering changing `vec` and `mat` to use `int` instead - they are designed to be _small_ vectors/matrices, so putting a large size_t makes little sense.
